File: ios/Fuel/Core/Logic/Pace.swift
import Foundation

// PURE Foundation port of frontend/src/app-editorial/pace.ts. Time-of-day
// calorie pacing: given what's been eaten and the daily goal, judges whether
// intake is tracking the part of the waking day that's elapsed — so the card
// can say "on pace" vs "ahead of pace" instead of only a remaining number.
//
// Named `CaloriePace` / `PaceStatus` / `PaceReading` to avoid colliding with
// the weight-loss `Pace` enum in Targets.swift.

enum PaceStatus: String, Sendable, Equatable {
  case over
  case ahead
  case on
  case behind
}

// i18n key for the human label (mirrors `PaceLabelKey` in the TS). The English
// copy matches lib/i18n/en.ts exactly; it is lowercase and title-cased at the
// render site.
enum PaceLabelKey: String, Sendable, Equatable {
  case overGoal = "pace.overGoal"
  case aheadOfPace = "pace.aheadOfPace"
  case roomToSpare = "pace.roomToSpare"
  case onPace = "pace.onPace"

  var text: String {
    switch self {
    case .overGoal: return "over goal"
    case .aheadOfPace: return "ahead of pace"
    case .roomToSpare: return "room to spare"
    case .onPace: return "on pace"
    }
  }
}

struct PaceReading: Sendable, Equatable {
  /// Goal minus consumed. Negative once over goal.
  let remaining: Int
  let status: PaceStatus
  let labelKey: PaceLabelKey

  var label: String { labelKey.text }
}

enum CaloriePace {
  // Assumed waking eating window (local hours). Pace is measured against how far
  // through this window we are, not the full 24h.
  static let wakeStart: Double = 8
  static let wakeEnd: Double = 22
  /// Tolerance around the expected-by-now figure before calling it off-pace.
  static let tolerance: Double = 0.08

  static func compute(
    consumed: Int,
    goal: Int,
    now: Date = Date(),
    calendar: Calendar = .current
  ) -> PaceReading {
    let remaining = goal - consumed

    // `over` is checked first — being past goal beats any time-of-day reasoning.
    if consumed > goal {
      return PaceReading(remaining: remaining, status: .over, labelKey: .overGoal)
    }

    let comps = calendar.dateComponents([.hour, .minute], from: now)
    let hour = Double(comps.hour ?? 0) + Double(comps.minute ?? 0) / 60
    let frac = min(max((hour - wakeStart) / (wakeEnd - wakeStart), 0), 1)
    let expected = Double(goal) * frac
    let tol = Double(goal) * tolerance
    let eaten = Double(consumed)

    if eaten > expected + tol {
      return PaceReading(remaining: remaining, status: .ahead, labelKey: .aheadOfPace)
    }
    if eaten < expected - tol {
      return PaceReading(remaining: remaining, status: .behind, labelKey: .roomToSpare)
    }
    return PaceReading(remaining: remaining, status: .on, labelKey: .onPace)
  }
}

File: frontend/src/app-editorial/pace.ts
/** Time-of-day calorie pacing. Given what's been eaten so far and the daily
 * goal, judges whether intake is tracking the part of the day that's elapsed —
 * so the card can say "on pace" vs "ahead of pace" instead of only showing a
 * remaining number that means nothing without knowing what time it is. */

// Assumed waking eating window (local hours). Pace is measured against how far
// through this window we are, not the full 24h — nobody eats evenly overnight.
const WAKE_START = 8
const WAKE_END = 22
/** Tolerance around the expected-by-now figure before calling it off-pace. */
const TOLERANCE = 0.08

export type PaceStatus = "over" | "ahead" | "on" | "behind"

// i18n key for the human label; translated at the render site (see
// macro-summary.tsx). Keeping the key here (not the string) so the util stays
// pure and the copy lives in the dictionaries.
export type PaceLabelKey =
  | "pace.overGoal"
  | "pace.aheadOfPace"
  | "pace.roomToSpare"
  | "pace.onPace"

export type Pace = {
  /** Goal minus consumed. Negative once over goal. */
  remaining: number
  status: PaceStatus
  labelKey: PaceLabelKey
}

export function computePace(consumed: number, goal: number, now: Date = new Date()): Pace {
  const remaining = goal - consumed

  if (consumed > goal) {
    return { remaining, status: "over", labelKey: "pace.overGoal" }
  }

  const hour = now.getHours() + now.getMinutes() / 60
  const frac = Math.min(Math.max((hour - WAKE_START) / (WAKE_END - WAKE_START), 0), 1)
  const expected = goal * frac
  const tol = goal * TOLERANCE

  if (consumed > expected + tol) return { remaining, status: "ahead", labelKey: "pace.aheadOfPace" }
  if (consumed < expected - tol) return { remaining, status: "behind", labelKey: "pace.roomToSpare" }
  return { remaining, status: "on", labelKey: "pace.onPace" }
}
